File: backend/src/services/bridges/alertEngineBridge.ts
/**
 * AlertEngine → EventBus 桥接
 *
 * 将 AlertEngine 触发的告警事件转换为 PerceptionEvent 注入 EventBus。
 * 作为内部告警感知源 (D1.5)。
 *
 * 工作方式：
 * - 注册为 EventBus 感知源
 * - 监听 AlertEngine 的 preprocessedEvent 回调
 * - 将 UnifiedEvent / CompositeEvent 转换为 type='alert' 的 PerceptionEvent
 */

import { logger } from '../../utils/logger';
import type { EventBus, Priority } from '../eventBus';
import type { AlertEngine } from '../ai-ops/alertEngine';
import type { UnifiedEvent, CompositeEvent, AlertSeverity } from '../../types/ai-ops';

export interface AlertEngineBridgeConfig {
  /** 是否启用，默认 true */
  enabled: boolean;
}

const DEFAULT_CONFIG: AlertEngineBridgeConfig = {
  enabled: true,
};

export class AlertEngineBridge {
  private readonly eventBus: EventBus;
  private readonly alertEngine: AlertEngine;
  private readonly config: AlertEngineBridgeConfig;
  private handler: ((event: UnifiedEvent | CompositeEvent) => void) | null = null;
  private started = false;

  constructor(
    eventBus: EventBus,
    alertEngine: AlertEngine,
    config?: Partial<AlertEngineBridgeConfig>,
  ) {
    this.eventBus = eventBus;
    this.alertEngine = alertEngine;
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * 启动桥接：注册感知源并监听 AlertEngine 事件
   */
  start(): void {
    if (this.started) return;
    if (!this.config.enabled) {
      logger.debug('[AlertEngineBridge] Disabled, skipping start');
      return;
    }

    // 注册为 EventBus 感知源 (D1.2)
    this.eventBus.registerSource({
      name: 'alert-engine-bridge',
      eventTypes: ['alert'],
      schemaVersion: '1.0.0',
    });

    // 监听 AlertEngine 的预处理事件回调
    this.handler = (event: UnifiedEvent | CompositeEvent) => {
      this.onPreprocessedEvent(event).catch((err) => {
        logger.warn('[AlertEngineBridge] Failed to publish alert event to EventBus', { error: err });
      });
    };

    this.alertEngine.onPreprocessedEvent(this.handler);
    this.started = true;
    logger.info('[AlertEngineBridge] Started');
  }

  /**
   * 停止桥接：取消监听
   */
  stop(): void {
    if (!this.started) return;
    if (this.handler) {
      this.alertEngine.offPreprocessedEvent(this.handler);
      this.handler = null;
    }
    this.started = false;
    logger.info('[AlertEngineBridge] Stopped');
  }

  // ─── 内部方法 ───

  private async onPreprocessedEvent(event: UnifiedEvent | CompositeEvent): Promise<void> {
    await this.publishAlertEvent(event);
  }

  private async publishAlertEvent(event: UnifiedEvent | CompositeEvent): Promise<void> {
    const priority = this.severityToPriority(event.severity);
    const isComposite = 'isComposite' in event && (event as CompositeEvent).isComposite;

    // 从 UnifiedEvent 的嵌套结构中提取字段
    const payload: Record<string, unknown> = {
      unifiedEventId: event.id,
      source: event.source,
      category: event.category,
      severity: event.severity,
      message: event.message,
      metadata: event.metadata,
    };

    // 告警规则信息（如果存在）
    if (event.alertRuleInfo) {
      payload.ruleId = event.alertRuleInfo.ruleId;
      payload.ruleName = event.alertRuleInfo.ruleName;
      payload.metric = event.alertRuleInfo.metric;
      payload.threshold = event.alertRuleInfo.threshold;
      payload.currentValue = event.alertRuleInfo.currentValue;
    }

    // 设备信息（如果存在）
    if (event.deviceInfo) {
      payload.deviceName = event.deviceInfo.hostname;
      payload.deviceIp = event.deviceInfo.ip;
    }

    // 复合事件聚合信息
    if (isComposite) {
      const composite = event as CompositeEvent;
      payload.isComposite = true;
      payload.childEvents = composite.childEvents;
      payload.aggregation = composite.aggregation;
    }

    await this.eventBus.publish({
      type: 'alert',
      priority,
      source: 'alert-engine-bridge',
      deviceId: event.deviceId,
      payload,
      schemaVersion: '1.0.0',
    });
  }

  private severityToPriority(severity: AlertSeverity | string): Priority {
    switch (severity) {
      case 'critical': return 'critical';
      case 'warning': return 'medium';
      case 'info': return 'info';
      default: return 'medium';
    }
  }
}

File: opsevo-python/src/opsevo/services/bridges/alert_engine_bridge.py
"""
AlertEngineBridge — AlertEngine → EventBus 桥接

将 AlertEngine 触发的告警事件转换为 PerceptionEvent 注入 EventBus。
"""

from __future__ import annotations

from typing import Any

import structlog

from opsevo.events.event_bus import EventBus
from opsevo.events.types import EventType, PerceptionEvent, Priority

logger = structlog.get_logger(__name__)


class AlertEngineBridge:
    """AlertEngine → EventBus 桥接器。"""

    def __init__(
        self,
        event_bus: EventBus,
        alert_engine: Any,
        enabled: bool = True,
    ) -> None:
        self._event_bus = event_bus
        self._alert_engine = alert_engine
        self._enabled = enabled
        self._started = False

    def start(self) -> None:
        if self._started:
            return
        if not self._enabled:
            logger.debug("AlertEngineBridge disabled")
            return
        self._event_bus.register_source(
            "alert-engine-bridge",
            {"event_types": ["alert"], "schema_version": "1.0.0"},
        )
        # 注册回调到 AlertEngine
        if hasattr(self._alert_engine, "on_preprocessed_event"):
            self._alert_engine.on_preprocessed_event(self._on_event)
        self._started = True
        logger.info("AlertEngineBridge started")

    def stop(self) -> None:
        if not self._started:
            return
        if hasattr(self._alert_engine, "off_preprocessed_event"):
            self._alert_engine.off_preprocessed_event(self._on_event)
        self._started = False
        logger.info("AlertEngineBridge stopped")

    async def _on_event(self, event: dict[str, Any]) -> None:
        try:
            await self._publish_alert(event)
        except Exception as exc:
            logger.warn("AlertEngineBridge publish failed", error=str(exc))

    async def _publish_alert(self, event: dict[str, Any]) -> None:
        severity = event.get("severity", "info")
        priority = self._severity_to_priority(severity)
        payload: dict[str, Any] = {
            "unified_event_id": event.get("id"),
            "source": event.get("source"),
            "category": event.get("category"),
            "severity": severity,
            "message": event.get("message"),
            "metadata": event.get("metadata"),
        }
        # 告警规则信息
        rule_info = event.get("alert_rule_info") or event.get("alertRuleInfo")
        if rule_info:
            payload["rule_id"] = rule_info.get("ruleId")
            payload["rule_name"] = rule_info.get("ruleName")
        # 设备信息
        device_info = event.get("device_info") or event.get("deviceInfo")
        if device_info:
            payload["device_name"] = device_info.get("hostname")
            payload["device_ip"] = device_info.get("ip")
        # 复合事件
        if event.get("is_composite") or event.get("isComposite"):
            payload["is_composite"] = True
            payload["child_events"] = event.get("child_events") or event.get("childEvents")

        perception = PerceptionEvent(
            type=EventType.ALERT,
            priority=priority,
            source="alert-engine-bridge",
            payload=payload,
            schema_version="1.0.0",
        )
        device_id = event.get("device_id") or event.get("deviceId")
        if device_id:
            perception.payload["device_id"] = device_id
        await self._event_bus.publish(perception)

    @staticmethod
    def _severity_to_priority(severity: str) -> Priority:
        mapping = {
            "critical": Priority.CRITICAL,
            "warning": Priority.MEDIUM,
            "info": Priority.INFO,
        }
        return mapping.get(severity, Priority.MEDIUM)
